spring-cloud-gateway-core/src/test/resources/application-client-auth-ssl.yml

```diff
@@ -0,0 +1,36 @@
+test:
+  uri: lb:https://testservice
+
+server:
+  ssl:
+    enabled: true
+    key-alias: scg
+    key-store-password: scg1234
+    key-store: classpath:scg-keystore.p12
+    trust-store: classpath:scg-truststore.p12
+    trust-store-password: scg1234
+    trust-store-type: PKCS12
+    key-store-type: PKCS12
+    client-auth: Need
+spring:
+  cloud:
+    gateway:
+      httpclient:
+        ssl:
```

Comment:
Would it be better to use a `key-store` and `trust-store` (like in `server.ssl`)? All the keys would be wrapped up in a more secure way then. Or is there a reason to prefer this way?

Reply:
Netty's SslContextBuilder provides following Apis for keyManager:
```
SslContextBuilder keyManager(java.io.File keyCertChainFile, java.io.File keyFile)
SslContextBuilder keyManager(java.io.InputStream keyCertChainInputStream, java.io.InputStream keyInputStream)
SslContextBuilder keyManager(java.security.PrivateKey key, java.security.cert.X509Certificate... keyCertChain)
SslContextBuilder keyManager(java.io.File keyCertChainFile, java.io.File keyFile, java.lang.String keyPassword)
SslContextBuilder keyManager(java.io.InputStream keyCertChainInputStream, java.io.InputStream keyInputStream, java.lang.String keyPassword)
SslContextBuilder keyManager(java.security.PrivateKey key, java.lang.String keyPassword, java.security.cert.X509Certificate... keyCertChain)
SslContextBuilder keyManager(javax.net.ssl.KeyManagerFactory keyManagerFactory)
```
And the key required in the Api's is a java.security.PrivateKey(in PKCS#8 format), if we use a keystore then have to include openssl code to extract PrivateKey and keycertChain from keystore(or some other option)

This will be overkill for the enhancement.

Reply:
In `org.springframework.boot.web.embedded.netty.SslServerCustomizer` there is some server side code that I think we should at least attempt to emulate. It uses `KeyManagerFactory` and also `ResourceUtils` to convert a `String` path to an `InputStream` where necessary. I think that we definitely want to avoid `java.io.File` directly if possible.

Reply:
See also here for more prior art.

https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/config/ClientHttpConnectorFactory.java#L56-L64

https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/support/SslConfiguration.java

It's possible we will see some features in Spring Framework and/or Spring Boot to support setting up SSL on netty clients. In lieu of that it might be better to go with a simple `Customizer` approach here and let users provide their own implementation until we have something more concrete.

Reply:
@dsyer  thanks for some interesting suggestions,
Removed File usage and implemented the same using keyManagerFactory
Also made changes in YAML to include path changes and case changes

Reply:
> I like this better now. Some of the documentation comments in the `HttpClientProperties` were a bit opaque, but that can easily be improved. Just one clarification: the X509 certs can be `Resource` locations as well (the only example we have is of a `File`)?

can you elaborate this a little?

Reply:
> See also here for more prior art.
> 
> https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/config/ClientHttpConnectorFactory.java#L56-L64
> 
> https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/main/java/org/springframework/vault/support/SslConfiguration.java
> 
> It's possible we will see some features in Spring Framework and/or Spring Boot to support setting up SSL on netty clients. In lieu of that it might be better to go with a simple `Customizer` approach here and let users provide their own implementation until we have something more concrete.

This is one good thing because of the fact that we keep onus of passing the sslConfig on client  ,
the spring-vault has it for keystore/truststore config only and even we're using same here.

Reply:
>  can you elaborate this a little?

I think it might be redundant? You replaced the certs config with a trust store?

Reply:
Truststore was needed for client-auth:Need, that's why added it

Reply:
Made corrections for logging and incorporated if inside HttpClientProperties in last commit